### PR TITLE
WIP: Smooth all the things

### PR DESCRIPTION
**Pull request by MayeulC**
Use bilinear interpolation for avatars and images in the timeline.

I'm still looking at smoothing the following:
* Room avatars (including ctrl+k)
* Communities avatars
* Pixellated avatars on profile views.

So there's probably another PR coming up shortly for these, though I have to dig a bit more into the code to find these (if you have a hint, I'm all ears, I've started to understand the caching and fetching mechanism, but the display one still eludes me a bit). I've had this itch for quite some time, but couldn't get to compile nheko. Might have been an LC issue?

## Before
partial_2020-08-21-210656_grim
(look at user avatars)

partial_2020-08-21-211422_grim
(my avatar is already using bilinear interpolation on the second screenshot, look at the picture)

## After
partial_2020-08-21-210817_grim
partial_2020-08-21-214547_grim

### CODE REVIEW

**Comment by MayeulC** [comment]
What is dynamic thumbnailing?

I guess this is also likely more noticeable on a low DPI (96) display like mine.

I am not sure how to benchmark performance. I guess I'll make a room with a lot of big images?

The property [could be disabled](https://wiki.qt.io/Performance_tip_Images#Enable_Image.smooth_only_when_necessary) during scrolling animation and resize).

Smoothing images in the timeline might come at a relatively big cost, depending on their size. I just assumed this wasn't something that would need to be done often. However, it might be worth subsampling big images before scaling them smoothly (I recall a blog post stating a O(n³) complexity advocating for just that, but I can't find it again).

----
Regarding fetched images, I think I would do the following:
* Only fetch a thumbnail that roughly corresponds to the desired resolution
  * Cache it, for every thumbnail size that hasn't been fetched yet
  * If that's a fixed-width UI element, smooth scale it, and store it at that size (qml property sourceSize)
* When some interaction requires a higher-size thumbnail:
  1. Serve the thumbnail from the cache, asynchronously fetch the resource in the background
  2. Replace the thumbnail with the newly fetched image

----

Something similar to [this kde patch](https://invent.kde.org/multimedia/kamoso/-/merge_requests/5) could also be done (I saw this in passing).

**Comment by MayeulC** [comment]
> They are then usually rendered by the src/ui/Avatar widget although the Community badges may be rendered differently in the src/CommunityListItem.cpp.

Found the first one, thank you. It looks like this patch 276df5d might have made c0ddb37 redundant, I'll see that later (tomorrow?). It makes such a huge difference, visually. And it had been looking the same on every computer I use.

I haven't looked at communities yet, there is also the room header.

276df5d somewhat fixed avatar in profile dialogs (they are a lot nicer to look at), but it looks like those avatars are upscaled from the timeline avatar, instead of downscaled from the original image. Something else I'll try to fix with that PR if that's easy enough.

File: resources/qml/delegates/ImageMessage.qml [hunk 1]
```diff
@@ -32,6 +32,8 @@ Item {
 		source: model.data.url.replace("mxc://", "image://MxcImage/")
 		asynchronous: true
 		fillMode: Image.PreserveAspectFit
+		smooth: true
+		mipmap: true
```
[the remark this answers is not in the document]
Comment (MayeulC) [reply resources/qml/delegates/ImageMessage.qml:36]: Actually, from the test I just made, only mipmap changes something. setting only smooth has the same effect as setting none.

I wanted to see if that was also the case with 276df5d if I was to remove mipmaps **and** smooth for room avatars in the timeline, but avatars remain smooth without both.

The reason you only set mipmaps to true for avatars might be the same only mipmaps has an effect here? (again, with 
276df5d, neither has an effect).

[the remark this answers is not in the document]
Comment (deepbluev7) [reply resources/qml/delegates/ImageMessage.qml:36]: Yeah, that would make sense, thank you for testing that!

**Comment by deepbluev7** [comment]
They are not downscaled from the original image, since we request a thumbnail of the 128x128 pixels. Homeservers should return the original image, if they don't have a thumbnail with at least that side. Synapse doesn't, but that is a bug: https://github.com/matrix-org/synapse/issues/1222. While we can work around that, I'm not really eager to do so, I'd rather fix the synapse bug.

Apart from that bug, thumbnails should already always be larger than requested, usually in those sizes: https://matrix.org/docs/spec/client_server/latest#thumbnails. With dynamic thumbnailing you get thumbnails always in the correct size, but you waste a bit of storage space.

I think dynamically enabling and disabling smoothing may actually cost us more performance, we would need to benchmark that :3

Also 276df5d and c0ddb37 are not redundant. They target different avatars. One is in Qml, the other in the old widget code.

**Pull request closed without merging**